### ISSUE TITLE
FEFaceEvaluation: efficiency of gather_evaluate + integrate_scatter

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#10921]
After implementing #10811 and the later re-structuring like #10904, I see that the Laplacian evaluated with `gather_evaluate` and `integrate_scatter` on the faces runs more slowly with `FE_DGQHermite` than with `FE_DGQ`, at least in 2D and with degree 3. This should not be the case because the Hermite case should access only half the vector data and also do somewhat fewer operations otherwise. We need to see over the performance before the next release. At least the gcc compiler generates way too many integer code; it might be that we need to force inlining for all lambdas we use.

**Comment by peterrum** [comment]
>  it might be that we need to force inlining for all lambdas we use.

I would have thought that the compiler inlines the code since the type of the lamdbas is templated.

**Comment by kronbichler** [comment]
In any case, we need to do a more comprehensive analysis of what instructions get generated and how to avoid them. I would like to finish #9794 first because that might also reveal something similar, so better address things only once.

**Comment by peterrum** [comment]
This is an important issue. But I would suggest to postpone. There will be additional modifications coming to make element-centric loops working for gradients and Hermite basis.

**Comment by peterrum** [comment]
@kronbichler If I remember `FEEvaluationData` helped here, didn't it?

**Comment by kronbichler** [comment]
It did definitely help, but we need to look at the assembly code at some point before the release, so we should keep this open.
